File: common/src/main/kotlin/org/virtuslab/inkuire/model/SerializableProjections.kt
package org.virtuslab.inkuire.model

sealed class SProjection

object SStar : SProjection()

sealed class SVariance<out T : SBound> : SProjection() {
    abstract val inner: T
}

data class SCovariance<out T : SBound>(override val inner: T) : SVariance<T>() {
    override fun toString() = "out"
}

data class SContravariance<out T : SBound>(override val inner: T) : SVariance<T>() {
    override fun toString() = "in"
}

data class SInvariance<out T : SBound>(override val inner: T) : SVariance<T>() {
    override fun toString() = ""
}

sealed class SBound : SProjection()

object SVoid : SBound()

object SJavaObject : SBound()

object SDynamic : SBound()

data class SUnresolvedBound(val name: String) : SBound()

data class STypeParameter(val dri: SDRI, val name: String) : SBound()

data class SNullable(val inner: SBound) : SBound()

data class SPrimitiveJavaType(val name: String) : SBound()

data class STypeConstructor(
    val dri: SDRI,
    val projections: List<SProjection>,
    val modifier: SFunctionModifiers = SFunctionModifiers.NONE
) : SBound()

enum class SFunctionModifiers {
    NONE, FUNCTION, EXTENSION
}

File: commonScala/shared/src/main/scala/org/virtuslab/inkuire/model/SerializableProjections.scala
package org.virtuslab.inkuire.model

sealed class SProjection

object SStar extends SProjection()

sealed abstract class SVariance extends SProjection() {
  val inner: SBound
}

case class SCovariance(override val inner: SBound) extends SVariance() {
  override def toString = "out"
}

case class SContravariance(override val inner: SBound) extends SVariance() {
  override def toString = "in"
}

case class SInvariance(override val inner: SBound) extends SVariance() {
  override def toString = ""
}

sealed class SBound extends SProjection()

object SVoid extends SBound()

object SJavaObject extends SBound()

object SDynamic extends SBound()

case class SUnresolvedBound(name: String) extends SBound()

case class STypeParameter(dri: SDRI, name: String) extends SBound()

case class SNullable(inner: SBound) extends SBound()

case class SPrimitiveJavaType(name: String) extends SBound()

case class STypeConstructor(
  dri:         SDRI,
  projections: List[SProjection],
  modifier:    SFunctionModifiers = SFunctionModifiers.NONE
) extends SBound()

sealed class SFunctionModifiers

object SFunctionModifiers {
  object NONE extends SFunctionModifiers
  object FUNCTION extends SFunctionModifiers
  object EXTENSION extends SFunctionModifiers
}
